src/System.Private.CoreLib/shared/System/Text/UTF8Encoding.Sealed.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Text
+{
+    public partial class UTF8Encoding
+    {
+        /// <summary>
+        /// A special instance of <see cref="UTF8Encoding"/> that is initialized with "don't throw on invalid sequences;
+        /// perform <see cref="Rune.ReplacementChar"/> substitution instead" semantics. This type allows for devirtualization
+        /// of calls made directly off of <see cref="Encoding.UTF8"/>. See https://github.com/dotnet/coreclr/pull/9230.
+        /// </summary>
+        internal sealed class UTF8EncodingSealed : UTF8Encoding
+        {
+            public UTF8EncodingSealed(bool encoderShouldEmitUTF8Identifier) : base(encoderShouldEmitUTF8Identifier) { }
+
+            public override ReadOnlySpan<byte> Preamble => _emitUTF8Identifier ? PreambleSpan : default;
+
+            public override byte[] GetBytes(string s)
+            {
+                // This method is short and can be inlined, meaning that the null check below
+                // might be elided if the JIT can prove not-null at the call site.
+
+                if (s != null && s.Length <= 32)
```

Comment:
```suggestion
                if (s?.Length <= 32)
```
Will result in the same IL.